eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ExternalDependencyExposedCheck.java

```diff
@@ -0,0 +1,201 @@
+     *
+     * @param token the IMPORT AST node
+     */
+    private void collectAllImportedClassPath(DetailAST token) {
+        final String importClassPath = FullIdent.createFullIdentBelow(token).getText();
+        final String className = importClassPath.substring(importClassPath.lastIndexOf(DOT) + 1);
+        classPathMap.put(className, importClassPath);
+    }
+
+    /**
+     * A Check for external dependency, log the error if it is an invalid external dependency.
+     *
+     * @param methodDefToken METHOD_DEF AST node
+     */
+    private void checkNoExternalDependencyExposed(DetailAST methodDefToken) {
+        final DetailAST modifiersToken = methodDefToken.findFirstToken(TokenTypes.MODIFIERS);
+        // Public API
+        if (modifiersToken != null &&
+            (modifiersToken.branchContains(TokenTypes.LITERAL_PUBLIC)
+                || modifiersToken.branchContains(TokenTypes.LITERAL_PROTECTED))) {
+            checkReturnType(methodDefToken.findFirstToken(TokenTypes.TYPE));
+            checkParametersType(methodDefToken.findFirstToken(TokenTypes.PARAMETERS));
+        }
+    }
+
+    /**
+     * Checks for return type
+     *
+     * @param typeToken TYPE AST node
+     */
+    private void checkReturnType(DetailAST typeToken) {
+        final int typeCount = typeToken.getChildCount();
+        // The TYPE node has more than one child, such as Map<T, T>
+        if (typeCount == 1) {
+            final String returnTypeName = typeToken.getFirstChild().getText();
+            // if not exist in the classPathMap, that implies the type is java default types, such as int.
+            if (classPathMap.containsKey(returnTypeName)) {
+                if (!isValidLibrary(classPathMap.get(returnTypeName))) {
+                    log(typeToken, String.format(EXTERNAL_DEPENDENCY_ERROR, returnTypeName));
+                }
+            }
+        } else {
+            checkTypeArguments(typeToken);
+        }
+    }
+
+    /**
+     * Checks for input parameters
+     *
+     * @param parametersTypeToken PARAMETERS AST node
+     */
+    private void checkParametersType(DetailAST parametersTypeToken) {
+        for (DetailAST ast = parametersTypeToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+            if (ast.getType() == TokenTypes.PARAMETER_DEF) {
+                checkTypeArguments(ast.findFirstToken(TokenTypes.TYPE));
+            }
+        }
+    }
+
+    /**
+     *  A helper function that checks TYPE AST node. Since both return type and input parameter argument type has
+     *  TYPE AST node under. This function applied to both.
+     *
+     * @param typeToken TYPE AST node
+     */
+    private void checkTypeArguments(DetailAST typeToken) {
+        final DetailAST identToken = typeToken.findFirstToken(TokenTypes.IDENT);
+        // if there is no IDENT token, implies the token is default java types.
+        if (identToken != null) {
```

Comment:
the 6 level nested if loops are really hard to read. I would suggest something like:
```java
if (identToken == null) {
   return;
}

// keep doing other stuff, because you know if you get to this point in the code, the identifier token is not null.
```